File: chaincode/supplychain.js
const shim = require('fabric-shim');
const util = require('util');

let SupplyChain = class {
    async Init(stub) {
        let productCounterBytes = await stub.getState('ProductCounterNo');
        if(!productCounterBytes) {
            let counter = 0
            try {
                await stub.putState('ProductCounterNo', Buffer.from(JSON.stringify(counter)));
            } catch {
                return shim.error('Failed to Initiate Product Counter');
            }
        }

        let orderCounterBytes = await stub.getState('OrderCounterNo');
        if(!orderCounterBytes) {
            let counter = 0
            try {
                await stub.putState('OrderCounterNo', Buffer.from(JSON.stringify(counter)));
            } catch {
                return shim.error('Failed to Initiate Order Counter');
            }
        }

        let userCounterBytes = await stub.getState('UserCounterNo');
        if(!userCounterBytes) {
            let counter = 0
            try {
                await stub.putState('UserCounterNo', Buffer.from(JSON.stringify(counter)));
            } catch {
                return shim.error('Failed to Initiate User Order Counter');
            }
        }
        console.info('=========== Instantiated SupplyChain Chaincode ===========');
        return shim.success();
    }

    async Invoke(stub) {
        let ret = stub.getFunctionAndParameters();
        let method = this[ret.fcn];
        if(!method) {
            console.log('No function of name: ' + + ret.fcn + ' found');
            return shim.success();
        }
        try {
            let payload = await method(stub, ret.params, this);
            return shim.success(payload);
        } catch (err) {
            console.log(err);
            return shim.error(err);
        }
    }

    async getCounter(stub, assetType) {
        let counterAsBytes = await stub.getState(assetType);
        return counterAsBytes;
    }

    async incrementCounter(stub, assetType) {
        let counterAsBytes = await stub.getState(assetType);
        counterAsBytes++;
        try {
            await stub.putState(assetType, counterAsBytes);
        } catch {
            throw new Error('Failed to Increment Counter');
        }
        return counterAsBytes;
    }

    async getTxTimestampChannel(stub) {
        try {
            return txTimeAsPtr = stub.getTxTimestamp();
        } catch (err) {
            throw new Error('Returning Error in TimeStamp');
        }
    }

    async initLedger(stub, args) {
        const entityUser = {
            name: 'admin',
            user_id: 'admin',
            email: 'mpdn99@google.com',
            user_type: 'admin',
            address: 'hanoi',
            password: 'adminpw'
        }
        try {
            stub.putState(entityUser.user_id, entityUser);
        } catch (error) {
            shim.error('Failed to create Entity Asset: ' + entityUser.user_id);
        }
        console.log('Added' + entityUser);
         return shim.success();
    }

    async signIn(stub, args) {
        try {
            let entityUserBytes = stub.getState(args[0]);
            if(entityUserBytes.password != args[1]) {
                return shim.error('Either id or password is wrong');
            }   
        } catch (error) {
            return shim.error('Cannot find Entity');
        }
        return shim.success(entityUserBytes);
    }

    async createUser(stub, args) {
        if(args.length != 5){
            return shim.error('Incorrect number of arguments, require 5 arguments');
        }

        if(args[0].length == 0){
            return shim.error('Name must be provided to register user');
        }

        if(args[1].length == 0){
            return shim.error('Email is mandatory');
        }

        if(args[2].length == 0){
            return shim.error('User type must be specified');
        }

        if(args[3].length == 0){
            return shim.error('Address must be non-empty');
        }

        if(args[4].length == 0){
            return shim.error('Password must be non-empty');
        }

        let userCounter = this.getCounter(stub, 'UserCounterNo');
        userCounter ++;

        let User = {
            name: args[0],
            user_id: 'User' + userCounter,
            email: args[1],
            user_type: args[2],
            address: args[3],
            password: args[4]
        }

        try {
            stub.putState(User.user_id, User);
            this.incrementCounter(stub, 'UserCounterNo');
        } catch (error) {
            return shim.error('Failed to register user: ' + User.user_id);
        }
        return shim.success(User);
    }

    async createProduct(stub, args){
        if(args.length != 3){
            return shim.error('Incorrect number of arguments, require 3 arguments');
        }

        if(args[0].length == 0){
            return shim.error('Name must be provided to create a product');
        }

        if(args[1].length == 0){
            return shim.error('Manufacturer_id must be provided');
        }

        if(args[2].length){
            return shim.error('Price must be non-empty');
        }

        userBytes = stub.getState(args[1]);
        if(!userBytes){
            return shim.error('Cannot find user');
        }
        if(userBytes.user_type != 'manufacturer'){
            return shim.error('User type must be manufacturer');
        }

        let productCounter = this.getCounter(stub, 'ProductCounterNo');
        productCounter ++;

        const txTimeAsPtr = this.getTxTimestampChannel(stub);
        let dates = {
            manufacturerDate: txTimeAsPtr
        }
        let comAsset = {
            product_id: 'Product' + productCounter,
            order_id: '',
            name: args[0],
            supplier_id: '',
            manufacturer: args[1],
            distributor: '',
            retailer: '',
            customer: '',
            status: 'Available',
            date: dates,
            price: args[2]
        }

        try {
            stub.putState(comAsset.product_id, comAsset);
            console.log('Success in creating Product Asset ' + comAsset);
        } catch (error) {
            return shim.error('Failed to create Product Asset: ' + comAsset.product_id);
        }
        return shim.success(comAsset);
    }

    async updateProduct(stub, args){
        if(args.length != 4){
            return shim.error('Incorrect number of arguments, required 4');
        }

        if(args[0].length == 0){
            return shim.error('Product Id must be provided');
        }

        if(args[1].length == 0){
            return shim.error('User Id must be provided');
        }

        if(args[2].length == 0){
            return shim.error('Product name must be provided');
        }

        if(args[3].length == 0){
            return shim.error('Product price must be provided');
        }

        userBytes = stub.getState(args[1]);
        if(!userBytes){
            return shim.error('Cannot find user');
        }

        if (userBytes.user_type == 'customer'){
            return shim.error('User type cannot be customer');
        }

        let productBytes = stub.getState(args[0])
        if(!productBytes){
            return shim.error('Cannot find Product');
        }
        productBytes.name = args[2];
        productBytes.price = args[3];

        try {
            stub.putState(productBytes.product_id, productBytes);
            console.log('Success in updating Product ' , + productBytes.product_id);
        } catch (error) {
            return shim.error('Failed to update product ' + productBytes.product_id);
        }
        return shim.success(productBytes);
    }

    async orderProduct(stub, args){
        if(args.length != 2){
            return shim.error("Incorrect number of arguments, required 2");
        }
        if(args[0].length != 0){
            return shim.error("Customer id must be provided");
        }
        if(args[1].length !=0){
            return shim.error("Product id must be provided");
        }
        let userBytes = stub.getState(args[0]);
        if(!userBytes){
            return shim.error('Cannot find customer');
        }
        if(userBytes.user_type != "customer"){
            return shim.error('User type must be customer');
        }
        let productBytes = stub.getState(args[1]);
        if(!productBytes){
            return shim.error('Cannot find product');
        }
        let orderCounter = this.getCounter(stub, "OrderCounterNo");
        orderCounter++;
        let txTimeAsPtr = this.getTxTimestampChannel(stub)
        productBytes.order_id = "Order" + orderCounter;
        productBytes.customer_id = userBytes.user_id;
        productBytes.status = "Ordered";
        productBytes.date.orderedDate = txTimeAsPtr;

        this.incrementCounter(stub, "OrderCounterNo");
        try {
            stub.putState(productBytes.product_id, productBytes);
        } catch (error) {
            return shim.error('Failed to place the order: ' + productBytes.product_id)
        }
        console.log('Order placed successfuly ' + productBytes.product_id)
        return shim.success(productBytes)
    }

    async deliveredProduct(stub, args){
        if(args.length != 1){
            return shim.error('Incorrect number of arguments, required 4');
        }
        if(args[0].length == 0){
            return shim.error('Product id must be provided');
        }
        let productBytes = stub.getState(args[0])
        if(!productBytes){
            return shim.error('Cannot find product');
        }
        if(productBytes.status != "Sold"){
            return shim.error('product is not delivered yet');
        }
        txTimeAsPtr = this.getTxTimestampChannel(stub);
        productBytes.date.deliverdDate = txTimeAsPtr;
        productBytes.status = 'Delivered'
        try {
            stub.putState(productBytes.product_id, productBytes);
        } catch (error) {
            return shim.error('Failed to update that product is delivered: ' + productBytes.product_id);
        }
        console.log('Success in delivering Product ' + productBytes.product_id);
        return shim.success(productBytes);
    }

    async sellToCustomer(stub, args){
        if(args.length != 1){
            return shim.error('Incorrect number of arguments, require');
        }

        if(args[0].length == 0){
            return shim.error('Product id must be provided');
        }
        let productBytes = stub.getState(args[0])
        if(!productBytes){
            return shim.error('Cannot find product');
        }
        if(productBytes.order_id == ""){
            return shim.error("Product has not been ordered yet");
        }
        if(productBytes.customer_id == ""){
            return shim.error('Customer Id should be set to sell to customer');
        }
        txTimeAsPtr = this.getTxTimestampChannel(stub);
        productBytes.data.sellToCustomerDate = txTimeAsPtr;
        productBytes.status = "Sold";
        try {
            stub.putState(productBytes.product_id, productBytes);
        } catch (error) {
            return shim.error('Failed to Sell to Customer: ' + productBytes.product_id);
        }
        console.log('Success in sending Product ' + productBytes.product_id)
        return shim.success(productBytes)
    }

    async sendToRetailer(stub, args){
        if(args.length != 2){
            return shim.error
        }
    }

    async queryAll(stub, args){
        if(args.length != 1){
            return shim.error('Incorect number of arguments, require 1');
        }
        if(args[0].length == 0){
            return shim.error('Asset type must be provided');
        }

    }
}

File: chaincode/supplychain.go
package main

import (
	"encoding/json"
	"fmt"
	"strconv"
	"time"

	"github.com/hyperledger/fabric-chaincode-go/pkg/cid"
	"github.com/hyperledger/fabric-contract-api-go/contractapi"
)

type Supplychain struct {
	contractapi.Contract
}

type CounterNO struct {
	Counter int `json:"counter"`
}

type User struct {
	Name         string `json:"Name`
	User_ID      string `json:"UserId"`
	Email        string `json:"Email"`
	UserType     string `json:"UserType"`
	Organization string `json:"Organization"`
	Address      string `json:"Address"`
	Password     string `json:"Password"`
}

type UserInfo struct {
	Name         string `json:"Name`
	User_ID      string `json:"UserId"`
	Email        string `json:"Email"`
	UserType     string `json:"UserType"`
	Organization string `json:"Organization"`
	Address      string `json:"Address"`
}

type ProductPosition struct {
	Date            string   `json:"Date"`
	Organization    string   `json:"Organization"`
	Longtitude string `json:"Longtitude"`
	Latitude   string `json:"Latitude"`
}

type Product struct {
	Product_ID      string            `json:"ProductId"`
	Name            string            `json:"Name"`
	Manufacturer_ID string            `json:"Manufacturer"`
	Distributor_ID  string            `json:"Distributor"`
	Retailer_ID     string            `json:"Retailer"`
	Consumer_ID     string            `json:"Consumer"`
	Status          string            `json:"Status"`
	Positions       []ProductPosition `json:"Position"`
	Price           float64           `json:"Price"`
}

func getCounter(ctx contractapi.TransactionContextInterface, AssetType string) int {
	counterJSON, _ := ctx.GetStub().GetState(AssetType)
	counter := CounterNO{}

	json.Unmarshal(counterJSON, &counter)
	return counter.Counter
}

func incrementCounter(ctx contractapi.TransactionContextInterface, AssetType string) int {
	counterJSON, _ := ctx.GetStub().GetState(AssetType)
	counter := CounterNO{}

	json.Unmarshal(counterJSON, &counter)
	counter.Counter++
	counterJSON, _ = json.Marshal(counter)

	err := ctx.GetStub().PutState(AssetType, counterJSON)
	if err != nil {
		fmt.Sprintf("Failed to Increment Counter")
	}
	return counter.Counter
}

func (t *Supplychain) GetTxTimestampChannel(ctx contractapi.TransactionContextInterface) (string, error) {
	txTimeAsPtr, err := ctx.GetStub().GetTxTimestamp()
	if err != nil {
		return "Error", err
	}
	timeStr := time.Unix(txTimeAsPtr.Seconds, int64(txTimeAsPtr.Nanos)).String()

	return timeStr, nil
}

func (t *Supplychain) InitLedger(ctx contractapi.TransactionContextInterface) error {

	//Init Manufacturer Admin Account
	userManufacturer := User{
		Name:         "Manufacturer Admin",
		User_ID:      "manufacturer-admin",
		UserType:     "admin",
		Organization: "ManufacturerOrg",
		Address:      "Hanoi",
		Password:     "admin@123",
	}

	userManufacturerJSON, err := json.Marshal(userManufacturer)
	if err != nil {
		return err
	}

	err = ctx.GetStub().PutState(userManufacturer.User_ID, userManufacturerJSON)
	if err != nil {
		return fmt.Errorf("Failed to put to world state. %s", err.Error())
	}

	//Init Distributor Admin Account
	userDistributor := User{
		Name:         "Distributor Admin",
		User_ID:      "distributor-admin",
		UserType:     "admin",
		Organization: "DistributorOrg",
		Address:      "Hanoi",
		Password:     "admin@123",
	}
	userDistributorJSON, err := json.Marshal(userDistributor)
	if err != nil {
		return err
	}

	err = ctx.GetStub().PutState(userDistributor.User_ID, userDistributorJSON)
	if err != nil {
		return fmt.Errorf("Failed to put to world state. %s", err.Error())
	}

	//Init Retailer Admin Account
	userRetailer := User{
		Name:         "Retailer Admin",
		User_ID:      "retailer-admin",
		UserType:     "admin",
		Organization: "RetailerOrg",
		Address:      "Hanoi",
		Password:     "admin@123",
	}
	userRetailerJSON, err := json.Marshal(userRetailer)
	if err != nil {
		return err
	}
	err = ctx.GetStub().PutState(userRetailer.User_ID, userRetailerJSON)
	if err != nil {
		return fmt.Errorf("Failed to put to world state. %s", err.Error())
	}

	return nil
}

func (t *Supplychain) SignIn(ctx contractapi.TransactionContextInterface, userID string, password string) (*User, error) {
	userJSON, err := ctx.GetStub().GetState(userID)
	if err != nil {
		return nil, fmt.Errorf("Failed to read from world state. %s", err.Error())
	}
	if userJSON == nil {
		return nil, fmt.Errorf("Cannot find User %s", userID)
	}

	user := UserInfo{}
	err = json.Unmarshal(userJSON, &user)
	if err != nil {
		return nil, err
	}

	if user.Password != password {
		return fmt.Errorf("Incorrect password")
	}

	return user, nil
}

func (t *Supplychain) CreateProduct(ctx contractapi.TransactionContextInterface, name string, manufacturerID string, longtitude string, latitude string, price string) error {
	//Authentication
	id, err := cid.GetMSPID(ctx.GetStub())
	if err != nil {
		return err
	}
	if id != "ManufacturerMSP" {
		return fmt.Errorf("User must be Manufacturer")
	}

	//Add product
	priceAsFloat, err := strconv.ParseFloat(price, 64)
	if err != nil {
		return fmt.Errorf("Failed to convert price: %s", err.Error())
	}

	productCounter := getCounter(ctx, "ProductCounterNO")
	productCounter++

	txTimeAsPtr, err := t.GetTxTimestampChannel(ctx)
	if err != nil {
		return fmt.Errorf("Error in Transaction Timestamp")
	}
	postion := ProductPosition{}
	postion.Date = txTimeAsPtr
	postion.Longtitude = longtitude
	postion.Latitude = latitude
	postion.Organization = manufacturerID

	product := Product{
		Product_ID:      "Product" + strconv.Itoa(productCounter),
		Name:            name,
		Manufacturer_ID: manufacturerID,
		Distributor_ID:  "",
		Retailer_ID:     "",
		Consumer_ID:     "",
		Status:          "Available",
		Positions:       []ProductPosition{postion},
		Price:           priceAsFloat,
	}
	productJSON, err := json.Marshal(product)
	if err != nil {
		return err
	}
	err = ctx.GetStub().PutState(product.Product_ID, productJSON)
	if err != nil {
		return fmt.Errorf("Failed to put to world state. %s", err.Error())
	}
	incrementCounter(ctx, "ProductCounterNO")

	return nil
}

func (t *Supplychain) UpdateProduct(ctx contractapi.TransactionContextInterface, productID string, name string, price string) error {
	//Authentication
	id, err := cid.GetMSPID(ctx.GetStub())
	if err != nil {
		return err
	}
	if id != "ManufacturerMSP" {
		return fmt.Errorf("User must be Manufacturer")
	}

	productJSON, err := ctx.GetStub().GetState(productID)
	if err != nil {
		return fmt.Errorf("Failed to read from world state. %s", err.Error())
	}
	if productJSON == nil {
		fmt.Errorf("Cannot find product %s", productID)
	}

	product := Product{}
	json.Unmarshal(productJSON, &product)

	if product.Distributor_ID != "" {
		return fmt.Errorf("Product has sent to Distributor. Cannot update")
	}

	//Update product
	priceAsFloat, err := strconv.ParseFloat(price, 64)
	if err != nil {
		return fmt.Errorf("Failed to convert price: %s", err.Error())
	}
	product.Name = name
	product.Price = priceAsFloat

	updatedProductJSON, err := json.Marshal(product)
	if err != nil {
		return err
	}
	ctx.GetStub().PutState(product.Product_ID, updatedProductJSON)

	return nil
}

func (t *Supplychain) SentToDistributor(ctx contractapi.TransactionContextInterface, productID string, distributorID string, longtitude string, latitude string) error {
	//Authentication
	id, err := cid.GetMSPID(ctx.GetStub())
	if err != nil {
		return err
	}
	if id != "DistributorMSP" {
		return fmt.Errorf("User must be Distributor")
	}

	productJSON, err := ctx.GetStub().GetState(productID)
	if err != nil {
		return fmt.Errorf("Failed to read from world state. %s", err.Error())
	}
	if productJSON == nil {
		fmt.Errorf("Cannot find product %s", productID)
	}

	product := Product{}
	json.Unmarshal(productJSON, &product)

	if product.Retailer_ID != "" {
		return fmt.Errorf("Product has sent to Retailer. Cannot update")
	}

	//Update product
	txTimeAsPtr, err := t.GetTxTimestampChannel(ctx)
	if err != nil {
		return fmt.Errorf("Error in Transaction Timestamp")
	}

	product.Distributor_ID = distributorID
	product.Positions = append(product.Positions, ProductPosition{Date: txTimeAsPtr, Latitude: latitude, Longtitude: longtitude , Organization: distributorID})

	updatedProductJSON, err := json.Marshal(product)
	if err != nil {
		return err
	}
	ctx.GetStub().PutState(product.Product_ID, updatedProductJSON)

	return nil
}

func (t *Supplychain) SentToRetailer(ctx contractapi.TransactionContextInterface, productID string, retailerID string, longtitude string, latitude string) error {
	//Authentication
	id, err := cid.GetMSPID(ctx.GetStub())
	if err != nil {
		return err
	}
	if id != "RetailerMSP" {
		return fmt.Errorf("User must be Retailer")
	}
	productJSON, err := ctx.GetStub().GetState(productID)
	if err != nil {
		return fmt.Errorf("Failed to read from world state. %s", err.Error())
	}
	if productJSON == nil {
		fmt.Errorf("Cannot find product %s", productID)
	}

	product := Product{}
	json.Unmarshal(productJSON, &product)

	if product.Consumer_ID != "" {
		return fmt.Errorf("Product has sent to Consumer. Cannot update")
	}

	//Update product
	txTimeAsPtr, err := t.GetTxTimestampChannel(ctx)
	if err != nil {
		return fmt.Errorf("Error in Transaction Timestamp")
	}

	product.Retailer_ID = retailerID
	product.Positions = append(product.Positions, ProductPosition{Date: txTimeAsPtr, Latitude: latitude, Longtitude: longtitude, Organization: retailerID})

	updatedProductJSON, err := json.Marshal(product)
	if err != nil {
		return err
	}
	ctx.GetStub().PutState(product.Product_ID, updatedProductJSON)

	return nil
}

func (t *Supplychain) SellToConsumer(ctx contractapi.TransactionContextInterface, productID string, consumerID string, longtitude string, latitude string) error {
	//Authentication
	id, err := cid.GetMSPID(ctx.GetStub())
	if err != nil {
		return err
	}
	if id != "RetailerMSP" {
		return fmt.Errorf("User must be Retailer")
	}

	//Update product
	productJSON, err := ctx.GetStub().GetState(productID)
	if err != nil {
		return fmt.Errorf("Failed to read from world state. %s", err.Error())
	}
	if productJSON == nil {
		fmt.Errorf("Cannot find product %s", productID)
	}

	product := Product{}
	json.Unmarshal(productJSON, &product)

	txTimeAsPtr, err := t.GetTxTimestampChannel(ctx)
	if err != nil {
		return fmt.Errorf("Error in Transaction Timestamp")
	}

	product.Consumer_ID = consumerID
	product.Positions = append(product.Positions, ProductPosition{Date: txTimeAsPtr, Latitude: latitude, Longtitude: longtitude, Organization: consumerID})
	product.Status = "Sold"

	updatedProductJSON, err := json.Marshal(product)
	if err != nil {
		return err
	}
	ctx.GetStub().PutState(product.Product_ID, updatedProductJSON)

	return nil
}

func (t *Supplychain) QueryProduct(ctx contractapi.TransactionContextInterface, productID string) (*Product, error) {
	productJSON, err := ctx.GetStub().GetState(productID)
	if err != nil {
		return nil, fmt.Errorf("Failed to read from world state. %s", err.Error())
	}

	if productJSON == nil {
		return nil, fmt.Errorf("%s does not exist", productID)
	}

	product := new(Product)
	err = json.Unmarshal(productJSON, product)
	if err != nil {
		return nil, err
	}
	return product, nil
}

func (t *Supplychain) QueryAllProducts(ctx contractapi.TransactionContextInterface) ([]*Product, error) {
	startKey := "PRODUCT0"
	endKey := "PRODUCT999"

	resultsIterator, err := ctx.GetStub().GetStateByRange(startKey, endKey)
	if err != nil {
		return nil, err
	}
	defer resultsIterator.Close()

	results := []*Product{}
	for resultsIterator.HasNext() {
		queryResponse, err := resultsIterator.Next()
		if err != nil {
			return nil, err
		}

		product := new(Product)
		_ = json.Unmarshal(queryResponse.Value, product)
		results = append(results, product)
	}

	return results, nil
}

func main() {
	chaincode, err := contractapi.NewChaincode(new(Supplychain))
	if err != nil {
		fmt.Printf("Error create supplychain chaincode: %s", err.Error())
		return
	}

	if err := chaincode.Start(); err != nil {
		fmt.Printf("Error starting supplychain chaincode: %s", err.Error())
	}
}
